### ISSUE TITLE
Migration broken from 0.2.19 (and older)

### ISSUE THREAD

**Issue by nijel** [python-social-auth/social-app-django#28]
This is followup for #16. 

With #25 the migration from 0.2.20 and 0.2.21 is working fine, however users upgrading from 0.2.19 and older will still end up with error.

It can end up with error described in #16:

```
KeyError: u"Migration social_django.0002_add_related_name dependencies reference nonexistent parent node (u'social_django', u'0001_initial')"
```

Or it can pop up as error while validating relations (see https://github.com/WeblateOrg/weblate/issues/1344):

```
ValueError: The field accounts.VerifiedEmail.social was declared with a lazy reference to 'social_django.usersocialauth', but app 'social_django' isn't installed.
```

The problem is that Django does not see the old migrations as they are just half applied (due to missing `social_auth` ones):

```
manage.py showmigrations
...
social_django
 [ ] 0005_auto_20160727_2333 (1 squashed migrations)
 [ ] 0006_partial
...
```

This is what is stored in the database:

id    | app        | name                           | applied                       
----- | ---------- | ------------------------------ | ------------------------------
70    | default    | 0001_initial                   | 2017-02-01 15:03:35.625824    
112   | default    | 0002_add_related_name          | 2017-02-01 15:03:41.684870    
113   | default    | 0003_alter_email_max_length    | 2017-02-01 15:03:41.739056    
114   | default    | 0004_auto_20160423_0400        | 2017-02-01 15:03:41.877357

**Comment by nijel** [comment]
The workaround that works is to upgrade to 0.2.21 first and then to social_django.

**Comment by yohanboniface** [comment]
More or less same issue here, when trying to migrate from python-social-auth==python-social-auth==0.2.14 to the new system.

```
Traceback (most recent call last):
  File "/home/umap/.virtualenvs/umap2/bin/umap", line 9, in <module>
    load_entry_point('umap==0.8.0.rc.13', 'console_scripts', 'umap')()
  File "/home/umap/.virtualenvs/umap2/lib/python3.4/site-packages/umap/bin/__init__.py", line 12, in main
    management.execute_from_command_line()
  File "/home/umap/.virtualenvs/umap2/lib/python3.4/site-packages/django/core/management/__init__.py", line 367, in execute_from_command_line
    utility.execute()
  File "/home/umap/.virtualenvs/umap2/lib/python3.4/site-packages/django/core/management/__init__.py", line 359, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/umap/.virtualenvs/umap2/lib/python3.4/site-packages/django/core/management/base.py", line 294, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/umap/.virtualenvs/umap2/lib/python3.4/site-packages/django/core/management/base.py", line 345, in execute
    output = self.handle(*args, **options)
  File "/home/umap/.virtualenvs/umap2/lib/python3.4/site-packages/django/core/management/commands/migrate.py", line 183, in handle
    executor.loader.project_state(),
  File "/home/umap/.virtualenvs/umap2/lib/python3.4/site-packages/django/db/migrations/loader.py", line 317, in project_state
    return self.graph.make_state(nodes=nodes, at_end=at_end, real_apps=list(self.unmigrated_apps))
  File "/home/umap/.virtualenvs/umap2/lib/python3.4/site-packages/django/db/migrations/graph.py", line 409, in make_state
    project_state = self.nodes[node].mutate_state(project_state, preserve=False)
  File "/home/umap/.virtualenvs/umap2/lib/python3.4/site-packages/django/db/migrations/migration.py", line 92, in mutate_state
    operation.state_forwards(self.app_label, new_state)
  File "/home/umap/.virtualenvs/umap2/lib/python3.4/site-packages/django/db/migrations/operations/fields.py", line 192, in state_forwards
    state.models[app_label, self.model_name_lower].fields
KeyError: ('social_django', 'usersocialauth')
```

```
django-admin showmigrations
…
social_django
 [ ] 0004_auto_20160423_0400 (2 squashed migrations)
 [ ] 0005_auto_20160727_2333 (1 squashed migrations)
 [ ] 0006_partial
…
```

**Comment by marcperrin** [comment]
Hi,
Sorry I'm not very knowledgeable about the "migrations" stuff - do you guys have a detailed, step-by-step guide to perform this workaround once I've run the "bad" migration and got this error?
Thanks!

**Comment by omab** [comment]
@nijel thanks for posting the results and the workaround, I've updated the migration guide to reflect this https://github.com/omab/python-social-auth/blob/master/MIGRATING_TO_SOCIAL.md#migrations.

@marcperrin take a look to the link above, that should help.

**Comment by coriolinus** [comment]
I'm encountering the same isssue, and the steps listed in the migrations guide do not help:

```shell
$ pip list --format=columns | grep social
python-social-auth        0.2.21
rest-social-auth          0.5.0
social-auth-app-django    1.1.0
social-auth-core          1.2.0
$ ./manage.py showmigrations social_django
social_django
 [ ] 0005_auto_20160727_2333 (1 squashed migrations)
 [ ] 0006_partial
$ ./manage.py migrate social_django
Operations to perform:
  Apply all migrations: social_django
Running migrations:
  Applying social_django.0005_auto_20160727_2333...Traceback (most recent call last):
  File "./manage.py", line 22, in <module>
    execute_from_command_line(sys.argv)
  File "/mnt/d/Users/coriolinus/Documents/Toptal/cardamom/.venv_wsl/lib/python3.5/site-packages/django/core/management/__init__.py", line 367, in execute_from_command_line
    utility.execute()
  File "/mnt/d/Users/coriolinus/Documents/Toptal/cardamom/.venv_wsl/lib/python3.5/site-packages/django/core/management/__init__.py", line 359, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/mnt/d/Users/coriolinus/Documents/Toptal/cardamom/.venv_wsl/lib/python3.5/site-packages/django/core/management/base.py", line 294, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/mnt/d/Users/coriolinus/Documents/Toptal/cardamom/.venv_wsl/lib/python3.5/site-packages/django/core/management/base.py", line 345, in execute
    output = self.handle(*args, **options)
  File "/mnt/d/Users/coriolinus/Documents/Toptal/cardamom/.venv_wsl/lib/python3.5/site-packages/django/core/management/commands/migrate.py", line 204, in handle
    fake_initial=fake_initial,
  File "/mnt/d/Users/coriolinus/Documents/Toptal/cardamom/.venv_wsl/lib/python3.5/site-packages/django/db/migrations/executor.py", line 115, in migrate
    state = self._migrate_all_forwards(state, plan, full_plan, fake=fake, fake_initial=fake_initial)
  File "/mnt/d/Users/coriolinus/Documents/Toptal/cardamom/.venv_wsl/lib/python3.5/site-packages/django/db/migrations/executor.py", line 145, in _migrate_all_forwards
    state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
  File "/mnt/d/Users/coriolinus/Documents/Toptal/cardamom/.venv_wsl/lib/python3.5/site-packages/django/db/migrations/executor.py", line 244, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/mnt/d/Users/coriolinus/Documents/Toptal/cardamom/.venv_wsl/lib/python3.5/site-packages/django/db/migrations/migration.py", line 119, in apply
    operation.state_forwards(self.app_label, project_state)
  File "/mnt/d/Users/coriolinus/Documents/Toptal/cardamom/.venv_wsl/lib/python3.5/site-packages/django/db/migrations/operations/models.py", line 521, in state_forwards
    model_state = state.models[app_label, self.name_lower]
KeyError: ('social_django', 'association')
```

**Comment by michi88** [comment]
I found that I had to uninstall python-social-auth and `rm -rf site-packages/social` because the uninstall leaves a migration file that conflicts. After that installing `0.2.21` and running the migrations worked for me.

**Comment by lukeaus** [comment]
I had a tricky time migrating, however I managed to overcome it using the following method to upgrade from python-social-auth v 0.2.19
1. upgrade to python-social-auth v 0.2.21

2. run ```python migrate social_auth```
Then your django_migrations table will look like this
```
default     | 0001_initial
default     | 0002_add_related_name
default     | 0003_alter_email_max_length
default     | 0004_auto_20160423_0400
social_auth | 0005_auto_20160727_2333
social_auth | 0001_initial
social_auth | 0003_alter_email_max_length
social_auth | 0004_auto_20160423_0400
social_auth | 0002_add_related_name
```

3. Run ```pip install social-auth-app-django>=1.0.0```
Use version >= 1.0.0 as I had some issues with v 0.0.1 - django was giving me a warning
```You have model changes without migrations blah....```

4. Update your project settings and imports under heading 'Settings' here https://github.com/omab/python-social-auth/blob/master/MIGRATING_TO_SOCIAL.md#migrations

5. Run all migrations ```python manage.py migrate'''

Now it should be working with all migrations applied.

**Comment by Kuein** [comment]
Any method for upgrade from django_social_auth==0.7.28 to social_django?
`Operations to perform:
  Apply all migrations: social_django
Running migrations:
  Applying social_django.0001_initial...Traceback (most recent call last):
  File "./manage.py", line 11, in <module>
    execute_from_command_line(sys.argv)
  File "/home/kuein/Projects/TogetherNetwork-Community/venv/lib/python2.7/site-packages/django/core/management/__init__.py", line 363, in execute_from_command_line
    utility.execute()
  File "/home/kuein/Projects/TogetherNetwork-Community/venv/lib/python2.7/site-packages/django/core/management/__init__.py", line 355, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/kuein/Projects/TogetherNetwork-Community/venv/lib/python2.7/site-packages/django/core/management/base.py", line 283, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/kuein/Projects/TogetherNetwork-Community/venv/lib/python2.7/site-packages/django/core/management/base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "/home/kuein/Projects/TogetherNetwork-Community/venv/lib/python2.7/site-packages/django/core/management/commands/migrate.py", line 204, in handle
    fake_initial=fake_initial,
  File "/home/kuein/Projects/TogetherNetwork-Community/venv/lib/python2.7/site-packages/django/db/migrations/executor.py", line 115, in migrate
    state = self._migrate_all_forwards(state, plan, full_plan, fake=fake, fake_initial=fake_initial)
  File "/home/kuein/Projects/TogetherNetwork-Community/venv/lib/python2.7/site-packages/django/db/migrations/executor.py", line 145, in _migrate_all_forwards
    state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
  File "/home/kuein/Projects/TogetherNetwork-Community/venv/lib/python2.7/site-packages/django/db/migrations/executor.py", line 244, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/home/kuein/Projects/TogetherNetwork-Community/venv/lib/python2.7/site-packages/django/db/migrations/migration.py", line 129, in apply
    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
  File "/home/kuein/Projects/TogetherNetwork-Community/venv/lib/python2.7/site-packages/django/db/migrations/operations/models.py", line 97, in database_forwards
    schema_editor.create_model(model)
  File "/home/kuein/Projects/TogetherNetwork-Community/venv/lib/python2.7/site-packages/django/db/backends/base/schema.py", line 302, in create_model
    self.execute(sql, params or None)
  File "/home/kuein/Projects/TogetherNetwork-Community/venv/lib/python2.7/site-packages/django/db/backends/base/schema.py", line 119, in execute
    cursor.execute(sql, params)
  File "/home/kuein/Projects/TogetherNetwork-Community/venv/lib/python2.7/site-packages/django/db/backends/utils.py", line 80, in execute
    return super(CursorDebugWrapper, self).execute(sql, params)
  File "/home/kuein/Projects/TogetherNetwork-Community/venv/lib/python2.7/site-packages/django/db/backends/utils.py", line 65, in execute
    return self.cursor.execute(sql, params)
  File "/home/kuein/Projects/TogetherNetwork-Community/venv/lib/python2.7/site-packages/django/db/utils.py", line 94, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/home/kuein/Projects/TogetherNetwork-Community/venv/lib/python2.7/site-packages/django/db/backends/utils.py", line 63, in execute
    return self.cursor.execute(sql)
  File "/home/kuein/Projects/TogetherNetwork-Community/venv/lib/python2.7/site-packages/django/db/backends/sqlite3/base.py", line 326, in execute
    return Database.Cursor.execute(self, query)
django.db.utils.OperationalError: table "social_auth_association" already exists
`

**Comment by rlfrahm** [comment]
@coriolinus I had that same issue after hammering (incorrectly) past the issue described in this comment: https://github.com/python-social-auth/social-app-django/issues/22#issuecomment-274850091

Doing the following db stuff got me past that issue. 

```bash
$ update django_migrations set app='social_django' where app='default';
```

Ref: https://github.com/python-social-auth/social-app-django/issues/22#issuecomment-274856118

Then I encountered your error @Kuein and did the following:

```bash
$ python manage.py migrate social_django 0001 --fake
```

then finish the migrations for `social_django`:

```bash
$ python manage.py migrate social_django
```

Then 🦄 and 🌈 appeared.. and everything works!

**Comment by imju** [comment]
One Caution
@rlfrahm the solution now should not include this line if you are using social-apps-django (1.2.0) 

`$ update django_migrations set app='social_django' where app='default';`

The migration checks both 'default' and 'social_django' in the migrations.  If you update this, then you will get 'NodeNotFoundError' in runserver.

**Comment by Spartach1985** [comment]
None of these methods worked for me when upgrading from Django 1.9 to 1.11: when trying to make `manage.py migrate`, I got a `KeyError: ('social_django', u'association')`.
That's what helped:

1. I dumped and then removed all four social_* tables from DB (only one of them - social_auth_usersocialauth - had some entries in it)
2. I made `pip uninstall` for all apps with 'social' in their name
3. `pip install social-auth-app-django`
4. `DELETE FROM django_migrations WHERE app='default';` (there were 4 such rows)
5. `python manage.py migrate`
6. I returned all dumped rows back to `social_auth_usersocialauth`. Lucky I am, its structure didn't change.
Not it looks as working app!

**Comment by marcperrin** [comment]
Échec de la remise pour ces destinataires ou groupes :

marcperrin (marc.perrin@dont-nod.com)<mailto:marc.perrin@dont-nod.com>
L'adresse de messagerie que vous avez entrée est introuvable. Vérifiez l'adresse de messagerie du destinataire et essayez de renvoyer le message. Si le problème persiste, contactez le support technique.







Informations de diagnostic pour les administrateurs :

Génération du serveur : dont-nod.com

marc.perrin@dont-nod.com
#550 5.1.1 RESOLVER.ADR.RecipNotFound; not found ##

En-têtes de message d'origine :

Received: from SRV (192.168.1.5) by SRV.dont-nod.lan (192.168.1.5) with
 Microsoft SMTP Server id 14.3.361.1; Wed, 18 Jul 2018 06:18:16 +0200
Return-Path: <noreply@github.com>
Received: from out-5.smtp.github.com ([192.30.252.196]) by mx.kundenserver.de
 (mxeue012 [212.227.15.41]) with ESMTPS (Nemesis) id 1MCrom-1foPLB1Zlw-009CwJ
 for <marc.perrin@dont-nod.com>; Wed, 18 Jul 2018 06:17:09 +0200
Date: Tue, 17 Jul 2018 21:17:08 -0700
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
        s=pf2014; t=1531887428;
        bh=PI1o+tHUiS0N/gEHScVwqXyjoOa+DIZT18PU1JjgGPs=;
        h=Date:From:Reply-To:To:Cc:In-Reply-To:References:Subject:List-ID:
         List-Archive:List-Post:List-Unsubscribe:From;
        b=xxTYFvWRhkW6ReKH1RDwH7lsAyCELh9hpfqe0mecZzXPiwQ7uCn1zrRkcGH4Nqj6v
         McSv6zsZsdcK0gyB6ixmf0n2acaYRd+xYhdC/yrwnZy3uFEnT5S0/aq2cVkoc+WusH
         7tOd1U9mXCR93PZikoHd/n4lJQFhSLxkabfhQxU4=

**Comment by CosmoRied** [comment]
Try apply the migrations backwards! - 
```

sudo heroku run python manage.py migrate social_django zero
 ›   Warning: heroku update available from 7.59.1 to 7.59.2.
Running python manage.py migrate social_django zero on ⬢ *************** up, run.2009 (Free)
/app/.heroku/python/lib/python3.7/site-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
  """)
Operations to perform:
  Unapply all migrations: social_django
Running migrations:
  Rendering model states... DONE
  Unapplying social_django.0010_uid_db_index... OK
  Unapplying social_django.0009_auto_20191118_0520... OK
  Unapplying social_django.0008_partial_timestamp... OK
  Unapplying social_django.0007_code_timestamp... OK
  Unapplying social_django.0006_partial... OK
  Unapplying social_django.0005_auto_20160727_2333... OK
  Unapplying social_django.0004_auto_20160423_0400... OK
  Unapplying social_django.0003_alter_email_max_length... OK
  Unapplying social_django.0002_add_related_name... OK
  Unapplying social_django.0001_initial...Traceback (most recent call last):

```

Then some error about the wrong number of constraints in the initial migration: 

`ValueError: Found wrong number (2) of constraints for social_auth_nonce(server_url, timestamp, salt)`

But when I tried to migrate again, no migrations needed to be applied.
